File: backend/src/galaxy.ts
import express from "express";
import toml from "smol-toml";
import multer, { Multer } from "multer";
import AdmZip from "adm-zip";
import { logger } from "./utils/logger.js";
import got, { Got } from "got";

export class Galaxy {
    private readonly upload: multer.Multer;
    got: Got;
    constructor() {
        this.upload = multer()
        this.got = got.extend({
            prefixUrl: "https://galaxy.tianpao.top/",
            //prefixUrl: "http://localhost:3000/",
            headers: {
                "User-Agent": "DeEarthX",
            },
            responseType: "json",
        });
    }
  getRouter() {
    const router = express.Router();
    router.use(express.json()); // 解析 JSON 请求体
    router.post("/upload",this.upload.array("files"), (req, res) => {
        const files = req.files as Express.Multer.File[];
        if(!files || files.length === 0){
            res.status(400).json({ status: 400, message: "No file uploaded" });
            return;
        }
        const modids = this.getModids(files);
        logger.info("Uploaded modids", modids);
        res.json({modids}).end();
    });
    router.post("/submit/:type",(req,res)=>{
        const type = req.params.type;
        if(type !== "server" && type !== "client"){
            res.status(400).json({ status: 400, message: "Invalid type" });
            return;
        }
        const modid = req.body.modids as string;
        if(!modid){
            res.status(400).json({ status: 400, message: "No modid provided" });
            return;
        }
        this.got.post(`api/mod/submit/${type}`,{
            json: {
                modid,
            }
        }).then((response)=>{
            logger.info(`Submitted modids for ${type}：`, response.body);
            res.json(response.body).end();
        }).catch((error)=>{
            logger.error(`Failed to submit modids for ${type}：`, error);
            res.status(500).json({ status: 500, message: "Failed to submit modids" });
        })
    })
    return router;
  }

  getModids(files:Express.Multer.File[]):string[] {
        let modid:string[] = [];
        for(const file of files){
            const zip = new AdmZip(file.buffer);
            const entries = zip.getEntries();
            for(const entry of entries){
                if(entry.entryName.endsWith("mods.toml")){
                    const content = entry.getData().toString("utf8");
                    const config = toml.parse(content) as any;
                    modid.push(config.mods[0].modId as string)
                }else if(entry.entryName.endsWith("neoforge.mods.toml")){
                    const content = entry.getData().toString("utf8");
                    const config = toml.parse(content) as any;
                    modid.push(config.mods[0].modId as string)
                }else if(entry.entryName.endsWith("fabric.mod.json")){
                    const content = entry.getData().toString("utf8");
                    const config = JSON.parse(content);
                    modid.push(config.id as string)
                }
        }
    }
        return modid
  }
}

File: core/galaxy/galaxy.go
package galaxy

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"

	"resty.dev/v3"

	"deearthx/core/util"
	"deearthx/core/ziputil"
)

// Galaxy handles mod ID submission to Galaxy Square
type Galaxy struct {
	client *resty.Client
}

// NewGalaxy creates a new Galaxy handler
func NewGalaxy() *Galaxy {
	return &Galaxy{
		client: resty.New().
			SetBaseURL("https://galaxy.tianpao.top/").
			SetHeader("User-Agent", "DeEarthX"),
	}
}

// UploadModsFromPaths extracts mod IDs from file paths
func (g *Galaxy) UploadModsFromPaths(paths []string) ([]string, error) {
	modIDs := []string{}

	for _, filePath := range paths {
		if _, err := os.Stat(filePath); err != nil {
			continue
		}

		data, err := os.ReadFile(filePath)
		if err != nil {
			continue
		}

		// Extract mod ID from JAR
		extractedIDs := g.extractModID(data)
		modIDs = append(modIDs, extractedIDs...)
	}

	util.Logger.Info("Extracted mod IDs from paths", "count", len(modIDs))
	return modIDs, nil
}

// UploadModsFromBuffers extracts mod IDs from file buffers
func (g *Galaxy) UploadModsFromBuffers(files [][]byte) ([]string, error) {
	modIDs := []string{}

	for _, data := range files {
		extractedIDs := g.extractModID(data)
		modIDs = append(modIDs, extractedIDs...)
	}

	return modIDs, nil
}

// SubmitModIDs submits mod IDs to Galaxy Square
func (g *Galaxy) SubmitModIDs(modType string, modIDs []string) error {
	modIDStr := strings.Join(modIDs, ",")

	resp, err := g.client.R().
		SetBody(map[string]string{"modid": modIDStr}).
		Post("mod/submit/" + modType)

	if err != nil {
		util.Logger.Error("Failed to submit mod IDs: " + err.Error())
		return err
	}

	if resp.StatusCode() >= 400 {
		util.Logger.Error("Galaxy API error: HTTP " + resp.Status())
		return nil
	}

	util.Logger.Info("Submitted mod IDs to Galaxy Square", "type", modType, "count", len(modIDs))
	return nil
}

func (g *Galaxy) extractModID(jarData []byte) []string {
	modIDs := []string{}

	entries, err := ziputil.ReadZip(jarData)
	if err != nil {
		return modIDs
	}

	for _, entry := range entries {
		// Forge/NeoForge mods.toml
		if strings.HasSuffix(entry.Name, "mods.toml") || strings.HasSuffix(entry.Name, "neoforge.mods.toml") {
			modID := parseForgeModID(string(entry.Data))
			if modID != "" {
				modIDs = append(modIDs, modID)
			}
		}

		// Fabric fabric.mod.json
		if strings.HasSuffix(entry.Name, "fabric.mod.json") {
			modID := parseFabricModID(string(entry.Data))
			if modID != "" {
				modIDs = append(modIDs, modID)
			}
		}
	}

	return modIDs
}

func parseForgeModID(tomlData string) string {
	// Simple TOML parsing for modId
	lines := strings.Split(tomlData, "\n")
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "modId") {
			parts := strings.SplitN(line, "=", 2)
			if len(parts) == 2 {
				modID := strings.Trim(strings.TrimSpace(parts[1]), "\"")
				return modID
			}
		}
	}
	return ""
}

func parseFabricModID(jsonData string) string {
	var config struct {
		ID string `json:"id"`
	}
	if err := json.Unmarshal([]byte(jsonData), &config); err != nil {
		return ""
	}
	return config.ID
}

// GetModIDsFromFile extracts mod IDs from a single file
func GetModIDsFromFile(filePath string) ([]string, error) {
	data, err := os.ReadFile(filePath)
	if err != nil {
		return nil, err
	}

	g := NewGalaxy()
	return g.extractModID(data), nil
}

// GetFilename returns just the filename from a path
func GetFilename(path string) string {
	return filepath.Base(path)
}